why-pandera/index.qmd

```diff
@@ -0,0 +1,240 @@
+
+## Considered options
+
+::: content-hidden
+List and describe some of the options, as well as some of the benefits
+and drawbacks for each option.
+:::
+
+-   [`frictionless-py`](https://framework.frictionlessdata.io/)
+-   [Pandera](https://pandera.readthedocs.io/en/stable/index.html)
+-   [Great
+    Expectations](https://docs.greatexpectations.io/docs/core/introduction/)
+-   [Pydantic](https://docs.pydantic.dev/latest/)
+
+### `frictionless-py`
+
+[`frictionless-py`](https://framework.frictionlessdata.io/) is the
+Python implementation of the Data Package standard by its parent
+organisation, and as such it would be the obvious choice for our use
+case. As well as functionality for data verification and validation, it
+supports checking metadata against the Data Package standard and
+building pipelines for transforming data.
+
+::::: columns
+::: column
+#### Benefits
+
+-   Supports both data verification and validation, although it is not
+    possible to run these checks separately.
+-   Multiple tabular data formats are supported, including `pandas` data
+    frames.
+-   Directly compatible with our JSON metadata, as it implements the
+    Data Package standard.
+-   Supports large data files.
+-   Supports extracting metadata from data, matching the Data Package
+    standard.
+:::
+
+::: column
+#### Drawbacks
+
+-   The API suggests that it is possible to filter for specific errors,
+    but this functionality does not seem to work fully.
+-   There are a number of different entry points to the
+    verification/validation flow and it is quite difficult to foresee
+    how these differ in behaviour.
+-   `polars` data frames are not supported.
+-   So far we've found it a bit difficult to navigate the
+    `frictionless-py` codebase and documentation.
+:::
+:::::
+
+### Pandera
+
+[Pandera](https://pandera.readthedocs.io/en/stable/index.html) is a
+flexible data validation library operating on data frames. Its
+validation mechanism is based on the concept of a schema expressing
+expectations about the data. It also has capabilities for preprocessing
+data and generating synthetic data from Pandera schemas.
+
+::::: columns
+::: column
+#### Benefits
+
+-   Supports both data verification and validation, and can run these
+    checks separately.
+-   Supports `polars` data frames to a large extent (see on the right).
+-   Supports large datasets.
+-   Offers schema inference, although not with `polars`.
+-   Pandera is widely used, extensively tested, and has good
+    documentation.
+:::
+
+::: column
+#### Drawbacks
+
+-   Only data frames are accepted as input, so other formats (e.g. CSV)
+    have to be loaded into a data frame first.
+-   While `polars` is supported, the integration is [not yet
+    complete](https://pandera.readthedocs.io/en/stable/index.html#supported-features).
+    E.g., it cannot yet extract metadata from `polars` data frames.
+-   We would need to write custom code to translate our table metadata
```

Comment:
With "table metadata" do you mean "resource metadata/properties"? Or specifically the table schema properties?

Reply:
Specifically table schema properties because here we care about what's inside the data file. So mostly the `fields` on the table schema and the `constraints` within.